Objects/typeobject.c

```diff
@@ -4747,6 +4877,82 @@ is_dunder_name(PyObject *name)
     return 0;
 }
 
+static void
+update_cache(struct type_cache_entry *entry, PyObject *name, unsigned int version_tag, PyObject *value)
+{
+    entry->version = version_tag;
+    entry->value = value;  /* borrowed */
+    assert(_PyASCIIObject_CAST(name)->hash != -1);
+    OBJECT_STAT_INC_COND(type_cache_collisions, entry->name != Py_None && entry->name != name);
+    Py_SETREF(entry->name, Py_NewRef(name));
+}
+
+#if Py_GIL_DISABLED
+
+#define TYPE_CACHE_IS_UPDATING(sequence) (sequence & 0x01)
+
+static void
+update_cache_gil_disabled(struct type_cache_entry *entry, PyObject *name,
+                          unsigned int version_tag, PyObject *value)
+{
+    // Similar to linux seqlock: https://en.wikipedia.org/wiki/Seqlock
+    // We use a sequence number to lock the writer, an even sequence means we're unlocked, an odd
+    // sequence means we're locked.
+    // Differs a little bit in that we use CAS on sequence as the lock, instead of a seperate spin lock.
+    // If our writer detects that another thread has already done the same write we'll just bail
+    // and restore the previous sequence number without doing any updates.
+
+    // lock the entry by setting by moving to an odd sequence number
+    int prev = entry->sequence;
+    while (1) {
+        if (TYPE_CACHE_IS_UPDATING(prev)) {
+            // Someone else is currently updating the cache
+            _Py_yield();
+            prev = _Py_atomic_load_int32_relaxed(&entry->sequence);
+        } else if(_Py_atomic_compare_exchange_int32(&entry->sequence, &prev, prev + 1)) {
+            // We've locked the cache
+            break;
+        } else {
+            _Py_yield();
+        }
+    }
```

Comment:
Can you refactor out the seqlock code into a type and functions, possibly in `pycore_lock.h`?